_posts/2020-07-19-rate-limiting-with-resilience4j.md

```diff
@@ -0,0 +1,291 @@
+Searching for flights; current time = 17:10:10 492
+...
+2020-07-27T17:10:10.494: Retry 'rateLimitedFlightSearch' recorded a successful retry attempt...
+[Flight{flightNumber='XY 765', flightDate='07/31/2020', from='NYC', to='LAX'}, ...]
+```
+
+The order in which we created the decorators is important. It would not work if we wrapped the `Retry` with the `RateLimiter`.
+
+## RateLimiter Events
+
+`RateLimiter` has an `EventPublisher` which generates events `RateLimiterOnSuccessEvent` and `RateLimiterOnFailureEvent` when calling a remote operation to indicate if acquiring permission was successful or not. We can listen for these events and log them, for example:
+
+```java
+RateLimiter limiter = registry.rateLimiter("flightSearchService");
+limiter.getEventPublisher().onSuccess(e -> System.out.println(e.toString()));
+limiter.getEventPublisher().onFailure(e -> System.out.println(e.toString()));
+```
+
+The sample output shows what's logged:
+
+```
+RateLimiterEvent{type=SUCCESSFUL_ACQUIRE, rateLimiterName='flightSearchService', creationTime=2020-07-21T19:14:33.127+05:30}
+... other lines omitted ...
+RateLimiterEvent{type=FAILED_ACQUIRE, rateLimiterName='flightSearchService', creationTime=2020-07-21T19:14:33.186+05:30}
+```
+
+## RateLimiter Metrics
+
+Suppose after implementing client-side throttling we find that the response times of our APIs have increased. This is possible because our request processing threads sometimes wait to get permission before calling the remote operation. How can we find out how often we are throttling our requests? How can we ensure that the values we've set on the `RateLimiterConfig` are working well?
+
+Monitoring `RateLimiter` metrics helps us identify such issues. If most of our request handling threads are often waiting to get permission, it could mean that our `limitForPeriod` is too low. Perhaps we need to work with our service provider and get additional quota provisioned first. 
+
+`RateLimiter` tracks two metrics: the number of permissions available (`resilience4j.ratelimiter.available.permissions`), and the number of threads waiting for permissions (`resilience4j.ratelimiter.waiting.threads`).
+
+First, we create `RateLimiterConfig`, `RateLimiterRegistry`, and `RateLimiter` as usual. Then, we create a `MeterRegistry` and bind the `RateLimiterRegistry` to it:
+
+```java
+MeterRegistry meterRegistry = new SimpleMeterRegistry();
+TaggedRateLimiterMetrics.ofRateLimiterRegistry(registry).bindTo(meterRegistry);
+```
+
+After running the rate-limited operation a few times, we display the captured metrics:
+
+```java
+Consumer<Meter> meterConsumer = meter -> {
+  String desc = meter.getId().getDescription();
+  String metricName = meter.getId().getName();
+  Double metricValue = StreamSupport.stream(meter.measure().spliterator(), false).
+    filter(m -> m.getStatistic().name().equals("VALUE")).
```

Comment:
I've never seen anyone format their code with the "." at the end of the line (even the code formatter in Jekyll had problems with that :). I've re-formatted it to put the "." at the start of the next line.

Reply:
Haha, it was a blind spot I guess. I had never noticed that. I had read somewhere that when you have long statements it's better to break the line after the operator so that a developer reading knowing that it continues, something like

int sum = a +
b +
...

I guess I had carried that over to "." as well. And you are right - I noticed after you mentioned that nearly everyone has the "." on the next line. I'll do the same next time. :)

Reply:
Right, I put operators like + and = on the same line, too.